sbnci/Modules/RecoEff/RecoEff_module.cc

```diff
@@ -0,0 +1,417 @@
+    fParticleTree->Branch("mc_yEnd",&mc_yEnd);
+    fParticleTree->Branch("mc_zEnd",&mc_zEnd);
+    fParticleTree->Branch("mc_pX0",&mc_pX0);
+    fParticleTree->Branch("mc_pY0",&mc_pY0);
+    fParticleTree->Branch("mc_pZ0",&mc_pZ0);
+    fParticleTree->Branch("mc_energy0",&mc_energy0);
+    fParticleTree->Branch("mc_momentum",&mc_momentum);
+    fParticleTree->Branch("mc_pXEnd",&mc_pXEnd);
+    fParticleTree->Branch("mc_pYEnd",&mc_pYEnd);
+    fParticleTree->Branch("mc_pZEnd",&mc_pZEnd);
+    fParticleTree->Branch("mc_energyEnd",&mc_energyEnd);
+    fParticleTree->Branch("mc_mass",&mc_mass);
+    fParticleTree->Branch("mc_theta_xy",&mc_theta_xy);
+    fParticleTree->Branch("mc_theta_yz",&mc_theta_yz);
+    fParticleTree->Branch("mc_theta_xz",&mc_theta_xz);
+    fParticleTree->Branch("mc_length",&mc_length);
+    fParticleTree->Branch("reco_isReconstructed",&reco_isReconstructed);
+    fParticleTree->Branch("reco_nTracks",&reco_nTracks);
+    fParticleTree->Branch("reco_nShowers",&reco_nShowers);
+    fParticleTree->Branch("reco_track_purity",&reco_track_purity);
+    fParticleTree->Branch("reco_track_completeness",&reco_track_completeness);
+    fParticleTree->Branch("reco_shower_purity",&reco_shower_purity);
+    fParticleTree->Branch("reco_shower_completeness",&reco_shower_completeness);
+    fParticleTree->Branch("reco_track_length",&reco_track_length);
+    fParticleTree->Branch("reco_shower_dEdx",&reco_shower_dEdx);
+  }
+
+void RecoEff::ClearData()
+{
+  n_tracks_map.clear(); track_comp_map.clear(); track_pur_map.clear();
+  n_showers_map.clear(); shower_comp_map.clear(); shower_pur_map.clear();
+  track_length_map.clear(); shower_dEdx_map.clear();
+  hits_map.clear();
+}
+
+void RecoEff::SetupHitsMap()
+{
+  for(unsigned hit_i = 0; hit_i < eHandleHits->size(); ++hit_i) {
+    const art::Ptr<recob::Hit> hit(eHandleHits,hit_i);
+    hits_map[TruthMatchUtils::TrueParticleID(clockData,hit,true)]++;
+  }
+}
+
+void RecoEff::ReconstructionProcessor(art::Event const &e)
+{
+  art::FindManyP<recob::Track> pfpTrackAssn(eHandlePFPs,e,fTrackModuleLabel);
+  art::FindManyP<recob::Shower> pfpShowerAssn(eHandlePFPs,e,fShowerModuleLabel);
+  art::FindManyP<recob::Hit> trackHitAssn(eHandleTracks,e,fTrackModuleLabel);
+  art::FindManyP<recob::Hit> showerHitAssn(eHandleShowers,e,fShowerModuleLabel);
+
+  const std::vector<art::Ptr<recob::PFParticle> > primaries = GetPrimaryPFPs();
+
+  for(auto primary : primaries){
+    const std::vector<long unsigned int> daughterIDs = primary->Daughters();
+
+    for(auto id: daughterIDs){
+      const art::Ptr<recob::PFParticle> daughter = GetPFP(id);
+      if(daughter.isNull()) continue;
+
+      if(daughter->PdgCode() == 13) {
+	const std::vector<art::Ptr<recob::Track> > tracksVec = pfpTrackAssn.at(daughter.key());
+	if(tracksVec.size() != 1) continue;
+	const art::Ptr<recob::Track> track = tracksVec[0];
+
+	float xOutEdge = 185, xInEdge = 1.5, yEdge = 185, zFrontEdge = 30, zBackEdge = 435;
```

Comment:
Can these be read from the geometry service.  The distance from the boundaries could be defined in the fcl files.  If we truly never want to change the values then maybe defining them as constexpr would improve efficiency a bit?

Reply:
Total drive-by (sorry). For simple constants like this `const` is fine. `constexpr` is new clever magic for cases that wouldn't be const under the old rules - eg because you call a (`constexpr`) function to initialize it. There is also `consteval` and `constinit`. This area got complicated... But here `const float` is fine. The const is less for efficiency (I strongly suspect it makes no difference), and more to document the fact that these aren't intended to be active variables in some calculation.

Reply:
I think the relevant difference between `const` and `constexpr` here is that the latter implies the value is known at compile time, which these numbers should be if they are hardcoded.
Defining a hardcoded values as constexpr, rather than const, also helps with readability (in my opinion).

Reply:
Hmm, you're right about the semantics, though in practice `const float x = <some constant>;` is certainly also going to initialize at compile time. My claim is that `constexpr` is sort of strange in the case where this is no "expression". I suppose it's a matter of taste. We have no end of style holy wars to fight [analogy that would have likely started one of those deleted].

Reply:
I think `constexpr` puts less reliance on the compiler to 'do the right thing' but, yeah, I am sure that it would be sensibly optimised.
Maybe `constexpr` use for const values is only just taking off but I feel like there are 'typical' cases already out there for using it in this way.  I dug this one out from brief googling: https://www.learncpp.com/cpp-tutorial/const-constexpr-and-symbolic-constants/
Ultimately, yes, this is often a stylistic thing.  I'd advocate that 'constexpr' readily implies the variable's intent in a way that 'const' does not.  I think you also get the 'static' ness for free...
 
I am becoming increasingly aware that we're debating what to do with a couple of floats...

Reply:
On the note of these being floats, @henrylay97, you're initialising your floats to be either ints or doubles. e.g. `float int xOutEdge = 185` vs `float xOutEdge = 185.f`

Reply:
I've gone with using Tom's geometry helper and fcl parameters, [its](https://github.com/SBNSoftware/sbnci/pull/7/commits/8e376db720dc2e8febc9a79c5cb565d1d3aacccc) not the most elegant because I was using different FVs for tracks and showers. It does allow it now to be varied from the fcl file though. This means I don't use either of the `const` options, sorry!